beacon_node/operation_pool/src/lib.rs

```diff
@@ -50,6 +66,97 @@ impl<T: EthSpec> OperationPool<T> {
         Self::default()
     }
 
+    /// Insert a sync contribution into the pool. We don't aggregate these contributions until they
+    /// are retrieved from the pool.
+    ///
+    /// ## Note
+    ///
+    /// This function assumes the given `contribution` is valid.
+    pub fn insert_sync_contribution(
+        &self,
+        contribution: SyncCommitteeContribution<T>,
+    ) -> Result<(), OpPoolError> {
+        let aggregate_id = SyncAggregateId::new(contribution.slot, contribution.beacon_block_root);
+        let mut contributions = self.sync_contributions.write();
+
+        match contributions.entry(aggregate_id) {
+            Entry::Vacant(entry) => {
+                // If no contributions exist for the key, insert the given contribution.
+                entry.insert(vec![contribution]);
+            }
+            Entry::Occupied(mut entry) => {
+                // If contributions exists for this key, check whether there exists a contribution
+                // with a matching `subcommittee_index`. If one exists, check whether the new or
+                // old contribution has more aggregation bits set. If the new one does, add it to the
+                // pool in place of the old one.
+                let existing_contributions = entry.get_mut();
+                match existing_contributions
+                    .iter_mut()
+                    .find(|existing_contribution| {
+                        existing_contribution.subcommittee_index == contribution.subcommittee_index
+                    }) {
+                    Some(existing_contribution) => {
+                        // Only need to replace the contribution if the new contribution has more
+                        // bits set.
+                        if existing_contribution.aggregation_bits.num_set_bits()
+                            < contribution.aggregation_bits.num_set_bits()
+                        {
+                            *existing_contribution = contribution;
```

Comment:
I notice that we don't every try to aggregate disjoint `SyncCommitteeContribution`s.

I.e., if `contribution` has a non-overlapping bitfield to `existing_contribution`, then add `contribution.signature` to `existing_contribution.signature` and also add the bitfields.

I'm just wondering if this has been discussed by @realbigsean or @michaelsproul before? I would be tempted to *not* include this for now, and potentially come back for it later.

Reply:
I was pretty conservative with my interpretation of this part of the spec:

> The proposer collects the contributions that match their local view of the chain (i.e. `contribution.beacon_block_root == block.parent_root`) for further aggregation when preparing a block. Of these contributions, proposers should select the best contribution seen across all aggregators for each subnet/subcommittee. A contribution with more valid signatures is better than a contribution with fewer signatures.

https://github.com/ethereum/eth2.0-specs/blob/v1.1.0-alpha.6/specs/altair/validator.md#sync-committee-1

Definitely agree it's something to keep in mind, but no, hadn't discussed it with michael before